File: almonds/app/stores/snippets.ts
import { defineStore } from "pinia";
import { invoke } from "@tauri-apps/api/core";

export interface Snippet {
  identifier: string;
  title: string | null;
  language: string | null;
  code: string;
  description: string | null;
  isPinned: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateSnippetPayload {
  title: string | null;
  language: string | null;
  code: string;
  description: string | null;
}

export type UpdateSnippetPayload = Partial<CreateSnippetPayload>;

export const useSnippetStore = defineStore("snippets_store", {
  state: () => ({
    snippets: [] as Snippet[],
    recent: [] as Snippet[],
    loading: false,
    recentLoading: false,
  }),

  getters: {
    languages: (state): string[] => {
      const langs = new Set<string>();
      for (const snippet of state.snippets) {
        if (snippet.language) langs.add(snippet.language);
      }
      return Array.from(langs).sort();
    },
  },

  actions: {
    async fetchSnippets() {
      this.loading = true;
      try {
        this.snippets = await invoke<Snippet[]>("get_all_snippets", {
          meta: await getWorkspaceMeta(),
        });
      } finally {
        this.loading = false;
      }
    },

    async fetchRecentSnippets() {
      this.recentLoading = true;
      try {
        this.recent = await invoke<Snippet[]>("get_recently_added_snippet", {
          meta: await getWorkspaceMeta(),
        });
      } finally {
        this.recentLoading = false;
      }
    },

    async createSnippet(payload: CreateSnippetPayload): Promise<Snippet> {
      const created = await invoke<Snippet>("create_snippet", {
        snippet: payload,
        meta: await getWorkspaceMeta(),
      });

      this.snippets.unshift(created);
      await this.fetchRecentSnippets();
      return created;
    },

    async updateSnippet(
      identifier: string,
      payload: UpdateSnippetPayload,
    ): Promise<Snippet> {
      const updated = await invoke<Snippet>("update_snippet", {
        identifier,
        snippet: payload,
        meta: await getWorkspaceMeta(),
      });

      const idx = this.snippets.findIndex((s) => s.identifier === identifier);
      if (idx !== -1) this.snippets[idx] = updated;

      return updated;
    },

    async deleteSnippet(identifier: string) {
      await invoke("delete_snippet", {
        identifier,
        meta: await getWorkspaceMeta(),
      });

      await this.fetchSnippets();
    },
    
    async duplicateSnippet(
      recordIdentifier: string,
      previousWorkspaceIdentifier: string,
      targetWorkspaceIdentifier: string,
    ) {
      await invoke("duplicate_snippet", {
        recordIdentifier,
        previousWorkspaceIdentifier,
        targetWorkspaceIdentifier,
        meta: await getWorkspaceMeta(),
      });
    
      await this.fetchSnippets();
    },
    
    async transferSnippet(
      recordIdentifier: string,
      previousWorkspaceIdentifier: string,
      targetWorkspaceIdentifier: string,
    ) {
      await invoke("transfer_snippet", {
        recordIdentifier,
        previousWorkspaceIdentifier,
        targetWorkspaceIdentifier,
        meta: await getWorkspaceMeta(),
      });
    
      this.snippets = this.snippets.filter(
        (s) => s.identifier !== recordIdentifier,
      );
    },
    
  },
});

File: shurbs/rust/src/commands/snippets.rs
use almond_kernel::{
    adapters::snippets::{CreateSnippet, UpdateSnippet},
    repositories::snippets::SnippetRepositoryExt,
    repositories::workspace_manager::{DuplicateRecord, TransferRecord},
};
use chrono::DateTime;

use crate::error::{make_meta, parse_uuid, AppError};
use crate::state::app_state;

#[flutter_rust_bridge::frb]
pub async fn create_snippet(
    code: String,
    title: Option<String>,
    language: Option<String>,
    description: Option<String>,
    is_pinned: bool,
    workspace_identifier: Option<String>,
    meta_workspace_id: Option<String>,
) -> Result<String, String> {
    let meta = make_meta(meta_workspace_id).map_err(|e| e.to_string())?;
    let ws_id = workspace_identifier
        .as_deref()
        .map(parse_uuid)
        .transpose()
        .map_err(|e: AppError| e.to_string())?;

    let now = DateTime::parse_from_rfc3339(&chrono::Utc::now().to_rfc3339()).unwrap();
    let payload = CreateSnippet {
        title,
        language,
        code,
        description,
        is_pinned,
        created_at: now,
        updated_at: now,
        workspace_identifier: ws_id,
    };

    let snippet = app_state()
        .snippets
        .create(&payload, &meta)
        .await
        .map_err(|e| e.to_string())?;

    serde_json::to_string(&snippet).map_err(|e| e.to_string())
}

#[flutter_rust_bridge::frb]
pub async fn get_snippet(
    identifier: String,
    meta_workspace_id: Option<String>,
) -> Result<Option<String>, String> {
    let id = parse_uuid(&identifier).map_err(|e| e.to_string())?;
    let meta = make_meta(meta_workspace_id).map_err(|e| e.to_string())?;

    let snippet = app_state()
        .snippets
        .find_by_id(&id, &meta)
        .await
        .map_err(|e| e.to_string())?;

    snippet
        .map(|s| serde_json::to_string(&s).map_err(|e| e.to_string()))
        .transpose()
}

#[flutter_rust_bridge::frb]
pub async fn get_all_snippets(meta_workspace_id: Option<String>) -> Result<String, String> {
    let meta = make_meta(meta_workspace_id).map_err(|e| e.to_string())?;
    let snippets = app_state()
        .snippets
        .find_all(&meta)
        .await
        .map_err(|e| e.to_string())?;

    serde_json::to_string(&snippets).map_err(|e| e.to_string())
}

#[flutter_rust_bridge::frb]
pub async fn get_recently_added_snippets(
    meta_workspace_id: Option<String>,
) -> Result<String, String> {
    let meta = make_meta(meta_workspace_id).map_err(|e| e.to_string())?;
    let snippets = app_state()
        .snippets
        .recently_added(&meta)
        .await
        .map_err(|e| e.to_string())?;

    serde_json::to_string(&snippets).map_err(|e| e.to_string())
}

#[flutter_rust_bridge::frb]
pub async fn update_snippet(
    identifier: String,
    title: Option<String>,
    language: Option<String>,
    code: Option<String>,
    description: Option<String>,
    is_pinned: Option<bool>,
    meta_workspace_id: Option<String>,
) -> Result<String, String> {
    let id = parse_uuid(&identifier).map_err(|e| e.to_string())?;
    let meta = make_meta(meta_workspace_id).map_err(|e| e.to_string())?;
    let payload = UpdateSnippet { title, language, code, description, is_pinned };

    let snippet = app_state()
        .snippets
        .update(&id, &payload, &meta)
        .await
        .map_err(|e| e.to_string())?;

    serde_json::to_string(&snippet).map_err(|e| e.to_string())
}

#[flutter_rust_bridge::frb]
pub async fn delete_snippet(
    identifier: String,
    meta_workspace_id: Option<String>,
) -> Result<(), String> {
    let id = parse_uuid(&identifier).map_err(|e| e.to_string())?;
    let meta = make_meta(meta_workspace_id).map_err(|e| e.to_string())?;
    app_state().snippets.delete(&id, &meta).await.map_err(|e| e.to_string())
}

#[flutter_rust_bridge::frb]
pub async fn duplicate_snippet(
    record_identifier: String,
    previous_workspace_identifier: String,
    target_workspace_identifier: String,
) -> Result<(), String> {
    let record = parse_uuid(&record_identifier).map_err(|e| e.to_string())?;
    let prev = parse_uuid(&previous_workspace_identifier).map_err(|e| e.to_string())?;
    let target = parse_uuid(&target_workspace_identifier).map_err(|e| e.to_string())?;

    app_state()
        .snippets
        .duplicate_record(&record, &prev, &target)
        .await
        .map_err(|e| e.to_string())
}

#[flutter_rust_bridge::frb]
pub async fn transfer_snippet(
    record_identifier: String,
    previous_workspace_identifier: String,
    target_workspace_identifier: String,
) -> Result<(), String> {
    let record = parse_uuid(&record_identifier).map_err(|e| e.to_string())?;
    let prev = parse_uuid(&previous_workspace_identifier).map_err(|e| e.to_string())?;
    let target = parse_uuid(&target_workspace_identifier).map_err(|e| e.to_string())?;

    app_state()
        .snippets
        .transfer_record(&record, &prev, &target)
        .await
        .map_err(|e| e.to_string())
}
